### PR TITLE
[SPARK-9551][SQL] add a cheap version of copy for UnsafeRow to reuse a copy buffer

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/sortBasedIterators.scala [hunk 1]
```diff
@@ -409,10 +409,19 @@ class FinalSortAggregationIterator(
...
+      copyBuffer = new Array[Byte](unsafeRow.getSizeInBytes)
+    }
+    unsafeRow.copyTo(copyBuffer)
+  }
+
   override def initialize(): Unit = {
     if (inputIter.hasNext) {
       initializeBuffer()
-      val currentRow = inputIter.next().copy()
+      val currentRow = cheapRowCopy(inputIter.next())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/sortBasedIterators.scala]: this is for demonstration, I'm not sure how to use this feature in aggregate sort.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 2]
```diff
@@ -442,6 +442,24 @@ public UnsafeRow copy() {
   }
 
   /**
+   * Copies this row, returning a UnsafeRow that stores its data to a user-provided byte[] buffer.
+   * User should ensure that provided buffer is large enough to store this row.
+   */
+  public UnsafeRow copyTo(byte[] buffer) {
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: i was thinking about this ... have:

a static method fromByteArray(numBytes: Int) that is a factory method

and then add a 

copyFrom(UnsafeRow other)

which basically copies the data from other into the current buffer.

If the size is too small, change the buffer to expand it. 

Then it is easy to use. Just call copyFrom where we call copy right now.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 3]
```diff
@@ -464,6 +464,37 @@ public UnsafeRow copy() {
   }
 
   /**
+   * Creates an empty UnsafeRow from a byte array with specified size.
+   * The returned row is invalid until we call copyFrom on it.
+   */
+  public static UnsafeRow createFromByteArray(int size) {
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: size -> numBytes

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 4]
```diff
@@ -464,6 +464,37 @@ public UnsafeRow copy() {
...
+    final UnsafeRow row = new UnsafeRow();
+    row.pointTo(new byte[size], PlatformDependent.BYTE_ARRAY_OFFSET, size);
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: i think the 2nd argument here is the number of fields, not offset

**Comment by rxin** [comment]
Can you also add unit test for these methods

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 3 above)
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: you might also need the number of fields

File: sql/core/src/test/scala/org/apache/spark/sql/UnsafeRowSuite.scala [hunk 5]
```diff
@@ -82,4 +82,42 @@ class UnsafeRowSuite extends SparkFunSuite {
...
+  test("createFromByteArray and copyFrom") {
+    val row = InternalRow(1, UTF8String.fromString("abc"))
+    val converter = UnsafeProjection.create(Array[DataType](IntegerType, StringType))
+    val unsafeRow = converter.apply(row)
+
+    val emptyRow = UnsafeRow.createFromByteArray(64, 2)
+    val buffer = emptyRow.getBaseObject
+
+    emptyRow.copyFrom(unsafeRow)
+    assert(emptyRow.getSizeInBytes() == unsafeRow.getSizeInBytes)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/UnsafeRowSuite.scala]: replace all == with ===

**Comment by rxin** [comment]
LGTM other than that.

**Comment by rxin** [comment]
Thanks - merging in master.

**Pull request closed without merging**